### DISCUSSION TITLE
Webhook for fetch API calls

### DISCUSSION THREAD

**Discussion by basavarajdodamani** [strapi/strapi#7352] [Legacy Topics]
Hello, 
Q1: Do we have an option to send event for a fetch call on the API. As i understand webhook is at the global level. 

Q2:I also would like to know if triggering of events can be controlled as per collection type. 

Q3: I somehow cant get below implementation working mentioned in: https://strapi.io/blog/strapi-webhooks 
I think URL mentioned in the custom.js is to be invoked when there is a post call on the collection type? 

A quick response on any of theses, is highly appreciated!

**Comment by alexandrebodin** [comment]
Hello @basavarajdodamani, this repository is for change Proposals only :) I'm moving your question to the main repository

**Comment by derrickmehaffy** [comment]
Hi @basavarajdodamani in this case I think it would probably be better to custom build your own webhook process based on the model lifecycles: https://strapi.io/documentation/v3.x/concepts/models.html#lifecycle-hooks

These are triggered when a `query` action happens (only the default queries are support, not custom ones), so in this case any time a service calls a `strapi.query()` action you will have two lifecycles:

- Before the query is fired
- After the query is fired

You can find more information on how to use these in the documentation I linked, but all queries are supported (the default ones)

- beforeFind/afterFind
- beforeFindOne/afterFindOne
- beforeCreate/afterCreate
- beforeUpdate/afterUpdate
- beforeDelete/afterDelete
- beforeCount/afterCount
- beforeSearch/afterSearch
- beforeCountSearch/afterCountSearch

**Reply by basavarajdodamani** [reply]
Hello @derrickmehaffy thank you for the quick response. beforeFind/afterFind work fine instead of afterFetch/BeforeFetch. This was helpful. Much appreciated. Noted on the custom ones. Any idea if this feature for custom routes is in the roadmap?

**Reply by derrickmehaffy** [reply]
Doesn't matter on custom routes, and I'm not sure about custom queries (I know for sure it won't happen with the RAW style such as with bookshelf and `strapi.connections.default.raw()` but I will ping @alexandrebodin about custom queries that tie into the bookshelf/mongoose methods. The hardest part from what I understand is we can't tell what the custom method is doing so you would have far less control.

I do know there is an RFC for some schema based lifecycles too from here: https://github.com/strapi/rfcs/pull/15

**Reply by alexandrebodin** [reply]
Right now we do not support other ways to trigger lifecycles that through the strapi.queries. You can however do it pretty easily in the services you create as it is custom code anyway
